### ISSUE TITLE
Moving up of custom configuration loading broke my setup (commit a28dbc8)

### ISSUE THREAD

**Issue by von** [robbyrussell/oh-my-zsh#4072]
Regarding commit a28dbc83937cbe624ec726199b4bbe7b65b2b2a0 which moved up the loading of $ZSH_CUSTOM/\*.zsh. My $ZSH_CUSTOM/*.zsh was relying on plugins and compinit having been called, so this commit completely breaks my configuration and it's not obvious to me how I can fix it.

If one relies on compinit and plugins in $ZSH_CUSTOM/*.zsh, what's the right course of action?

**Comment by gabetax** [comment]
This also broke my customizations that require plugins to already be loaded.

**Comment by gabetax** [comment]
I'm using customizations to replace a something from the `git` plugin, and this commit also broke it for me:

```sh
unalias g
g() { git ${@:-status} }
```

The "proper" way to do this (according to README) is to fork the plugin itself into the custom directory, but maintaining forks is a pain.

/cc @MatthewHager @ncanceill

**Comment by apjanke** [comment]
Caused by #3617 (which was included in #3809).

I'm not currently using custom files to tweak plugins like this, but have done it in the past. It would be nice to have it work, because yeah, forking plugins is a pain.

**Comment by qiuwei** [comment]
After this change, even customized keybinding stops working. 
My ^R is mysteriously bound to redisplay.

**Comment by ncanceill** [comment]
I think this was discussed somewhere else already, but searching is uneasy on a smartphone.

The way I remember it is to have a custom plugin instead of a simple file.

For instance you can override the git plugin completely by using ˋ$ZSH_CUSTOM/plugins/git/git.plugin.zshˋ.

Alternatively you can load something on top of it, in order to selectively override stuff, by having ˋ$ZSH_CUSTOM/plugins/gitcustom/gitcustom.plugin.zshˋ and adding ˋgitcustomˋ after ˋgitˋ to your ˋpluginsˋ array.

Not sure how well this is documented.

**Comment by apjanke** [comment]
It's not documented at all: the [existing Customization wiki page](https://github.com/robbyrussell/oh-my-zsh/wiki/Customization/4108a7c747d9b6586d90c1e01d26c6d9a4b5a54b) explicitly said that user custom files are loaded at the very end of OMZ initialization, and can be used to customize plugins by overriding individual bits. (Which was still wrong, since the theme was loaded after the custom code.) This other "patch" plugin mechanism wasn't mentioned at all.

I've updated the [Design](https://github.com/robbyrussell/oh-my-zsh/wiki/Design) and [Customization](https://github.com/robbyrussell/oh-my-zsh/wiki/Customization) wiki pages to reflect the new load order behavior and expand on how customization works for partial and complete overriding of libs and plugins. New section [Partially overriding an existing plugin](https://github.com/robbyrussell/oh-my-zsh/wiki/Customization#partially-overriding-an-existing-plugin) documents this technique, calling it a "patch plugin".

The README customization section could use a link to the more detailed Wiki page. (Although I would like to see more detailed documentation in the distribution itself, so it could be version-controlled and updated as part of any PRs that change its behavior.)

@qiuwei: maybe one of the plugins you load is clobbering your keybindings, now that plugins run after your custom code that would have had your custom key bindings.

**Comment by apjanke** [comment]
It seems like there's a case for customization hooks at both points in the load sequence: the new early point for stuff you want plugins to use at load time, and the old late point for partially customizing plugins after load.

As a hack, you can get an equivalent to the old behavior by sticking your custom code in a custom plugin, and putting that at the end of the `$plugins` load order. (And before the change, you could have gotten the new behavior as a hack by putting a custom plugin at the begining of `$plugins`.)

I think it would be nicer to have a named entry point instead of having to use a `$plugins` hack. Intent would be more clear, you could edit `$plugins` more easily without worrying about including your patch, and it might be more robust with respect to possible future changes to `oh-my-zsh.sh` initialization order. Philosophically cleaner too, IMHO. Maybe we could have a "post" or "late" hook, where everything in `$ZSH_CUSTOM/post/*.zsh` gets loaded at the old load point, after plugins are loaded.

**Comment by von** [comment]
Similar to what @apjanke suggested, I implemented a quick and dirty loading of plugins before and after customization hooks.

I create a customization file $ZSH_CUSTOM/001-preload.zsh (see https://gist.github.com/von/5892d439455d68701c83 for contents) named specifically to be loaded first. Basically it duplicates parts of oh-my-zsh.sh (it was created almost entirely by cut-n-paste):

1. Adds plugins to fpath
2. Calls 'autoload -U compinit' 
3. Loads plugins from plugins
4. Sets plugins=() so that oh-my-zsh.sh doesn't reload the plugins set in zshrc and subsequent custom files can add to it such that oh-my-zsh will then subsequently load those plugins

I have not tested part 4 by setting plugins in my custom files, but it elegantly fixes things such that oh-my-zsh works for me again. The 'autoload -U compinit is autoloaded twice, but this doesn't seem to cause any issue for me.

It would seem one could maintain backward compatibility to duplicating the plugin loading in oh-my-zsh.sh before and after the loading of customization hooks. Obviously one could put some of the code into functions to avoid duplicate code. I'm not sure where the 'autoload -U compinit' belongs. If it were up to me I'd move it up before the load of customization hooks, but I don't now if the movement of the customization hooks before it was intentionally or just a byproduct.

**Comment by apjanke** [comment]
The movement of customization hooks was intentional: that was the whole reason for #3617, so the custom files could define things, including redefining `lib/*.zsh` internals, which were used by plugins at plugin load time.

If you have two plugin loading points, how do you decide which plugin gets loaded when? This basically seems like it just restores the old behavior of the custom code being loaded after plugins. Since this leaves `plugins=()`, the second plugin loading pass is always a no-op. (And then `$plugins` in the shell doesn't reflect what was actually loaded by OMZ.)

**Comment by von** [comment]
> If you have two plugin loading points, how do you decide which plugin gets loaded when?

If you want a plugin loaded prior to the customization hooks, add it to plugins in zshrc. If you want it loaded after the customization hooks, add it to plugins in a customization hook (presumably where you set values that impact it).

> (And then $plugins in the shell doesn't reflect what was actually loaded by OMZ.)

Agreed. If that's desired, some additional logic is required.

**Comment by sudowork** [comment]
Just adding another data point. This also ended up breaking my configuration in a similar way - I was relying on the custom *.zsh files being loaded after plugins in order to override aliases, bindkeys, etc. My current workaround to this is to have a very small custom plugin that loads everything in `$ZSH_CUSTOM/after/*.zsh`. Then I just moved the relevant customizations into that directory and added the plugin to the end of my plugins to load.

**Comment by von** [comment]
It seems clear to me at this point that this commit (a28dbc8) made a non-trivial change to oh-my-zsh's interface with negative impacts that weren't anticipated. Can someone familiar with how the project makes decisions chime in with how this issue will get resolved? E.g. will this change get reverted? Backward compatibility added? Or is this a "won't fix"? Thanks.

**Comment by apjanke** [comment]
I think @robbyrussell, @mcornella, or @ncanceill would be the ones to speak to that.

**Comment by apjanke** [comment]
My vote would be to revert this change, and to add a new, separate hook for running custom code at the new location before the plugins and completion initialization. (e.g. `$ZSH_CUSTOM/early/*.zsh`.) That would fix existing user code, and still support the new pre-plugin functionality people wanted.

Reasoning: This change did break existing functionality that was out in the wild, changed core behavior which was explicitly documented, and we've heard from some users that doing the fix to work with the new way is difficult. Since this requires work on the user's part to fix, and there's no automatic prompt for them to do so or explain how, we're going to keep getting additional "my `compdef` is broken" reports from users, and it's going to take maintainer time to continue dealing with those. 

There may also be users out there whose custom code is no longer working, and they haven't reported it, because they don't know how, or don't even know what went wrong. (E.g. if it's a key binding thing and some of their arrow keys just stopped working one day, it may not be obvious exactly what's broken or where to look to fix it.)

I also find the "make your code a plugin and put it at the end of the list" workaround unsatisfactory: it feels a little ugly, it's prone to load-order interaction with other plugins down the line. And user code may break if the plugin design changes. E.g. if we change so that plugins are loaded inside a function call, so broken top-level `local` calls in the plugin files are fixed.

I do not think it's possible to add backward compatibility without essentially reverting this. It's a basic change to the initialization order. Any new mechanism to allow custom code to load at a different location would have to either require user intervention (in which case we're in the same situation as now) or default to the old behavior.

**Comment by BrendanThompson** [comment]
I +1 the suggestion there by @apjanke seems to be a lot of stuff with dependencies on having the files load after.

**Comment by mcornella** [comment]
Hi people, sorry for the late response. I submitted PR #4230 which reverts this commit. 

I've been thinking that perhaps a better solution would be to load custom files before plugins but *after* core lib files, as well as running `compinit` right at the very start of the init script (after modifying `$fpath` to look for completion files):

1. Modify `$fpath` based on enabled plugins.
2. Run `compinit`
3. Load core libraries.
4. Load custom files.
5. Load plugins.
6. Set theme.

What do you think?

**Comment by von** [comment]
@mcornella I use some plugins in my custom files so -1 from me on your suggested change.

**Comment by apjanke** [comment]
+1 on the PR to revert, but I think the new suggested ordering would just run in to the same problem again. Doing the custom files anywhere except the original location of "after plugins and before theme" would cause people breakage, because people have existing order-dependent code that relies on that exact positioning. And it doesn't accomplish the original goal of #3617, which was allowing people's custom code to specify new plugins to load, which must be done before the `$fpath`/`compinit` setup to be fully correct.

I'd suggest this load order instead:

1. Run "early" custom code from `$ZSH_CUSTOM/early/*.zsh`
2. Modify `$fpath` based on enabled plugins.
3. Run `compinit`.
4. Load core libraries.
5. Load plugins.
6. Load custom files. (from `$ZSH_CUSTOM/*.zsh`)
7. Set theme.

It's the same as the current load order, but introduces a new hook at the beginning of the startup sequence for people to run code early enough to configure plugins.

**Comment by MatthewHager** [comment]
We have a ton of stuff that will break if we aren't able to load and customize plugins in custom files. :(

**Comment by apjanke** [comment]
And as you can see in this bug report, other people have stuff that will break if you are able to. It sounds like there are incompatible goals here, and someone's going to have to be inconvenienced. Is one group more numerous or important than the other?

Unless there's a real strong case for it, I think reverting and adding a new hook sounds like the better choice. This was longstanding, documented core behavior that is an interface to code outside the OMZ repo. Changing that's kind of a big deal, and we don't really know how many people were affected by it. It might be better to err on the side of stability and back-compatibility. I know this is making me feel a little less confident about OMZ.

**Comment by MatthewHager** [comment]
Since we've got two factions going on this, maybe we should discuss the pros and cons of the new way. The pros with the new way are that the system is more extensible in a cleaner way. The cons are that legacy code will need to be updated to work with it.

I'm thinking we should focus on if we think the new way is good and if so, if we have the people update the legacy code and we all move forward. If we revert, we are stuck with a system in which you can't easily extend in certain ways.

**Comment by imanel** [comment]
I love how term "legacy code" solves this problem for you...

No - it just broke thousands of people configurations, making some functions not work, other throw errors, and another group of failing silently. What is worse in many cases it's so confusing that it's hard to debug - in my case all my aliases suddenly stopped working, and later I figured out they are simply overwritten by plugins (which never happened in past). I'm pretty sure there is HUGE number of people that didn't found this Github issue yet and trying to figure out WTF happened since last update. My solution was to revert to old git revision and disable updates "until this will be fixed".

There is a term - it's called "backward compatibility". And another: "public API". They are usually nearby - mostly because people are using them and expect to be stable. If, after several years, public API is changed with absolutely no reason then expect people to be angry.

My suggestion is as follows: revert to old behavior (it's not "legacy code", it's "working as specified in documentation for years"), and add new hook. There is exactly 0 arguments agains it - "new ways is more extensible" is great argument for creating new hook and leaving old one.

**Comment by von** [comment]
+1 To @imanel and @apjanke 

To me this seems straight forward. You need a hook in a different place, add a new hook in that place, don't move one that other people count on. I'm seen no one arguing against either hook, it's just removing the "legacy" one that causes problems.

**Comment by apjanke** [comment]
Totally. And it's an easy change; [here's an implementation](https://github.com/apjanke/oh-my-zsh/tree/early-custom-hook). 

If we do that, then MatthewHager's group can just move their custom `$ZSH_CUSTOM/*.zsh` files to `$ZSH_CUSTOM/early/*.zsh` and they'll work how they want, and everybody else's existing stuff will continue to work. (Including the poor users who were affected by this and already modified their code to use the late-loading custom plugin hack since we've let this sit for a month and a half already. And except the people who've already written code relying on the new #3617 way. Hope there's not too many of them.)

**Comment by apjanke** [comment]
I generally agree with @imanel. This will also affect the OMZ maintainers: some of those users trying to figure out what's going on will identify it as an OMZ problem, but not tie it to this issue directly. So I think you'll see a trickle of bug reports like "Comption not working on Debian" that will take OMZ volunteer time to diagnose, identify as this thing, and tell users how to fix. And the new way will take more time to document and explain how to use. OMZ's maintainers are stretched pretty thin: there's >600 open issues/PRs already; we don't need them arriving faster if we can help it.

**Comment by MatthewHager** [comment]
So long as the before hook has an env variable where you can override the location like you can with the current custom folder, that works for me. I recommend using before instead of early since that has been the pattern in other areas.

**Comment by apjanke** [comment]
Which other areas? I don't know other OMZ code that does much with load ordering. I picked "early" because it's relatively early in the loading sequence, but not before the entire sequence, so "pre" or "before" sounded a little extreme. (E.g. it's still after the auto-update check, initial modification to `$fpath`, and initialization of `$ZSH_CUSTOM` and `$ZSH_CACHE_DIR`.)

That example implementation respects `$ZSH_CUSTOM`, like the other custom code support in `oh-my-zsh.sh`.
